email-encryption/src/com/axonivy/market/emailencryption/service/EncryptedEmailSender.java

```diff
@@ -0,0 +1,207 @@
+import org.eclipse.core.resources.IProject;
+
+import ch.ivyteam.ivy.environment.Ivy;
+import ch.ivyteam.ivy.mail.MailClientConfig;
+import ch.ivyteam.ivy.mail.MailClientConfigProvider;
+import ch.ivyteam.ivy.mail.MailConstants.EmailEncryption;
+import ch.ivyteam.ivy.project.IIvyProject;
+import ch.ivyteam.ivy.scripting.objects.File;
+import ch.ivyteam.ivy.security.SecurityManagerFactory;
+import ch.ivyteam.util.IvyException;
+
+/**
+ * Service to send encrypted mails.
+ * 
+ */
+public class EncryptedEmailSender {
+
+	/**
+	 * Sends email encrypted by given certificate to provided receiver.
+	 * 
+	 * @param mail
+	 * @param certificate
+	 * @throws Exception
+	 */
+	public static void sendEncryptedMail(Email mail, X509Certificate certificate) throws Exception {
+		if (mail == null || StringUtils.isBlank(mail.getTo()) || certificate == null) {
+			throw new IvyException("Invalid mail settings for encrypted mail sending.");
+		}
+
+		ClassLoader csBackup = Thread.currentThread().getContextClassLoader();
+		Thread.currentThread().setContextClassLoader(EncryptedEmailSender.class.getClassLoader());
+
+		MailcapCommandMap mc = (MailcapCommandMap) CommandMap.getDefaultCommandMap();
+		mc.addMailcap(
+				"multipart/*;;x-java-content-handler=com.sun.mail.handlers.multipart_mixed;x-java-fallback-entry=true");
+		CommandMap.setDefaultCommandMap(mc);
+
+		MimeMultipart bodyMultipart = new MimeMultipart();
+
+		MimeBodyPart bodyMessagePart = new MimeBodyPart();
+		bodyMessagePart.setText(mail.getContent(), "UTF-8", "html");
+		bodyMultipart.addBodyPart(bodyMessagePart);
+
+		for (File attachment: mail.getAttachments()) {
+			try {
+				addAttachment(bodyMultipart, attachment);
+			} catch (UnsupportedEncodingException e) {
+				Ivy.log().error(e.toString());
+			}
+		}
+		// Get a Session object and create the mail message
+		Properties props = null;
+		Session session = null;
+
+		try {
+			MailClientConfig emailConfig = mailSetup();
+			props = getProperties(emailConfig);
+			session = getSession(props, emailConfig);
+		} catch (Exception e) {
+			throw new IvyException(e);
+		}
+
+		if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
+			Security.insertProviderAt(new BouncyCastleProvider(), 1);
+		}
+
+		SMIMEEnvelopedGenerator gen = new SMIMEEnvelopedGenerator();
+		JceKeyTransRecipientInfoGenerator keyGen = new JceKeyTransRecipientInfoGenerator(certificate);
+		keyGen.setProvider(BouncyCastleProvider.PROVIDER_NAME);
+		gen.addRecipientInfoGenerator(keyGen);
+		OutputEncryptor encryptor = new JceCMSContentEncryptorBuilder(CMSAlgorithm.RC2_CBC)
+				.setProvider(BouncyCastleProvider.PROVIDER_NAME).build();
+
+		MimeMessage messageToEncrypt = new MimeMessage(session);
+		messageToEncrypt.setContent(bodyMultipart);
+		messageToEncrypt.saveChanges();
+		MimeBodyPart contentPart = gen.generate(messageToEncrypt, encryptor);
+
+		Address fromUser = new InternetAddress(StringUtils.remove(mail.getFrom(), ";"));

```

Comment:
we had problems with UTF-8 addresses in the past. not sure if this is properly working... ?!